lib/methods/ifwapichecktask.cpp

```diff
@@ -0,0 +1,460 @@
+	if (resolvedMacros && !useResolvedMacros) {
+		auto commandEndpoint (checkable->GetCommandEndpoint());
+
+		// There's indeed a command endpoint, obviously for the actual check execution
+		if (commandEndpoint) {
+			// But it doesn't have this function, yet ("ifw-api-check-command")
+			if (!(commandEndpoint->GetCapabilities() & (uint_fast64_t)ApiCapabilities::IfwApiCheckCommand)) {
+				// Assume "ifw-api-check-command" has been imported into a check command which can also work
+				// based on "plugin-check-command", delegate respectively and hope for the best
+				PluginCheckTask::ScriptFunc(checkable, cr, resolvedMacros, useResolvedMacros);
+				return;
+			}
+		}
+	}
+
+	CheckCommand::Ptr command = CheckCommand::ExecuteOverride ? CheckCommand::ExecuteOverride : checkable->GetCheckCommand();
+	auto lcr (checkable->GetLastCheckResult());
+
+	Host::Ptr host;
+	Service::Ptr service;
+	tie(host, service) = GetHostService(checkable);
+
+	MacroProcessor::ResolverList resolvers;
+
+	if (MacroResolver::OverrideMacros)
+		resolvers.emplace_back("override", MacroResolver::OverrideMacros);
+
+	if (service)
+		resolvers.emplace_back("service", service);
+	resolvers.emplace_back("host", host);
+	resolvers.emplace_back("command", command);
+
+	auto resolveMacros ([&resolvers, &lcr, &resolvedMacros, useResolvedMacros](const char* macros) -> Value {
+		return MacroProcessor::ResolveMacros(
+			macros, resolvers, lcr, nullptr, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros
+		);
+	});
+
+	String psCommand = resolveMacros("$ifw_api_command$");
+	Dictionary::Ptr arguments = resolveMacros("$ifw_api_arguments$");
+	String psHost = resolveMacros("$ifw_api_host$");
+	String psPort = resolveMacros("$ifw_api_port$");
+	String expectedSan = resolveMacros("$ifw_api_expected_san$");
+	String cert = resolveMacros("$ifw_api_cert$");
+	String key = resolveMacros("$ifw_api_key$");
+	String ca = resolveMacros("$ifw_api_ca$");
+	String crl = resolveMacros("$ifw_api_crl$");
+	String username = resolveMacros("$ifw_api_username$");
+	String password = resolveMacros("$ifw_api_password$");
+
+	Dictionary::Ptr params = new Dictionary();
+
+	if (arguments) {
+		ObjectLock oLock (arguments);
+		Array::Ptr emptyCmd = new Array();
+
+		for (auto& kv : arguments) {
+			Dictionary::Ptr argSpec;
+
+			if (kv.second.IsObjectType<Dictionary>()) {
+				argSpec = Dictionary::Ptr(kv.second)->ShallowClone();
+			} else {
+				argSpec = new Dictionary({{ "value", kv.second }});
+			}
+
+			// See default branch of below switch
+			argSpec->Set("repeat_key", false);
+
+			{
+				ObjectLock oLock (argSpec);
+
+				for (auto& kv : argSpec) {
+					if (kv.second.IsObjectType<Function>()) {
+						auto now (Utility::GetTime());
+
+						ReportIfwCheckResult(
+							checkable, command->GetName(), cr,
+							"$ifw_api_arguments$ may not directly contain functions.", now, now
+						);
+
+						return;
+					}
+				}
```

Comment:
The check looks quite lax. I mean yes, this should catch the most likely cases, but after all, anything that's not properly serializable to JSON will behave in a strange way.

We're building something new, so we should be able to be more strict here. Something like "must be a map from string to trivial value".

Now I'm going a step back and asking myself, how much of the original argument handling is still left? Have a look at https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#checkcommand-arguments and quite a few of these things wouldn't work anyways:
- `skip_key`: I don't think anything good would happen if this was set, the code would just assume the first value to be the key and do something unexpected I guess.
- `order`: pointless for `ifw-api` as already discussed (https://github.com/Icinga/icinga-powershell-framework/pull/633#issuecomment-1580733732, https://github.com/Icinga/icinga-powershell-framework/pull/633#issuecomment-1580805573).
- `repeat_key`: pointless to set as it's already forced to `false` by the implementation.
- `key`: not sure if this could also be used for something useful with `ifw-api`, but I guess if you have to different arguments with `key` set to the same string, one would overwrite the other (instead of both being appended like it would happen when generating an actual command line).
- `separator`: sounds like another one that would just result in breakage if it was set, unlikely that you want `{"-flag=value": true}` in the JSON.

On the other hand, https://github.com/Icinga/icinga-powershell-framework/pull/633 currently only uses `value` and `set_if` without lambdas (because it can't anyways). This makes me wonder I we could radically simplify something here. What if we made `ifw_api_arguments` just a dictionary with strings as keys and (macro) strings as values, nothing else. Something like this:

```
vars.ifw_api_arguments = {
  "-WarnThreshold" = "$my_check_warn$"
  "-EnableDebug" = "$my_check_debug$"
}
```

Yes, this does not differentiate between `value` and `set_if`, but in the end, a JSON object is generated where for flags, you add `true` as value anyways.

The upside of such a change should be that the following code could be simplified quite a lot:
https://github.com/Icinga/icinga2/blob/249bccf49a531f66546aabe5cc2f9c9864bc0c61/lib/methods/ifwapichecktask.cpp#L290-L352

And it should boil down to something like this:
```c++
if (!kv.second.IsString()) {
  // error out
}
Value val = resolveMacros(kv.second);

if (!val.IsEmpty()) {
  params->Set(kv.first, val);
}
```

Maybe there's some tweaking of the second if required (what happens with `""` and `false` for example). And would also need a cross-check that this works fine with all types that exist in IfW checks.

Spinning that thought a bit further, why did we discard the idea of just having a variable that evaluates to the body, so that we don't hard-code any assumptions into C++ at all? I think that was because that function might be too complex, but that was at a point where we thought we could just reuse `command.arguments`. Now, I think it might even be as simple as that:

```
vars.ifw_api_body = {{ Json.encode({
  "-WarnThreshold" = macro("$my_check_warn$")
  "-EnableDebug" = macro("$my_check_debug$")
}) }}
```

(Now that I've written all of this, how the hell did I get from "The check looks quite lax." to "What do you thing of this entirely different suggestion?" :sweat_smile:)

Reply:
<\/walloftext>

1. Functions are the only type which is both already supported in Command#arguments and not serialisable. Lax or not, it is just exactly what we need. 
2. Look at how the macro processor interprets set_if. You can set `true`, `"true"`, `"1"`, `1`, ... Duplicating such magic logic is a bad idea in the long term. In contrast re-using both the code and the data structure LGTM.

Also, didn’t you worry about future planning in the first place? What if @LordHepipud will like/have to introduce something with both value and set_if?

> * `skip_key`: I don't think anything good would happen if this was set, the code would just assume the first value to be the key and do something ~unexpected~ I guess.

Now as you mention it, this usage is completely valid:

https://github.com/Icinga/icinga-powershell-framework/blob/e8c55d37d445c00c45d6f3f1090e437906a9cddd/lib/core/tools/Get-IcingaCheckCommandConfig.psm1#L102-L115

Reply:
Also there's no final decision yet whether we already need both value and set_if, see Icinga/icinga-powershell-framework#633 / "Things to consider while reviewing".

Reply:
> 1. Functions are the only type which is both already supported in Command#arguments and not serialisable. Lax or not, it is just exactly what we need.

Would there be any problem if the check did exclude all object types and not just functions? Everything that's listed in https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#checkcommand-arguments that's not a function is also not an object (as in what's covered by `ValueEnum` in `enum ValueType`).

> 2. Look at how the macro processor interprets set_if. You can set `true`, `"true"`, `"1"`, `1`, ... Duplicating such magic logic is a bad idea in the long term. In contrast re-using both the code and the data structure LGTM.

My point is that what this PR is currently doing is not fully compatible with what `command.arguments` as used in `PluginCheckTask`, see the list in my previous comment. And some of them are just fundamental differences between JSON objects and command line strings, so you couldn't even make it compatible some time in the future by implementing the missing pieces.

So do we want to do something complex that's like half-compatible, or do we want to see if a super simple solution would work as well.

> Also, didn’t you worry about future planning in the first place? What if @LordHepipud will like/have to introduce something with both value and set_if?

Yes, if we'd do this, this has to be coordinated. Just as it's the case for the current state of `ifw_api_arguments`.

> > * `skip_key`: I don't think anything good would happen if this was set, the code would just assume the first value to be the key and do something ~unexpected~ I guess.
> 
> Now as you mention it, this usage is completely valid:
> 
> https://github.com/Icinga/icinga-powershell-framework/blob/e8c55d37d445c00c45d6f3f1090e437906a9cddd/lib/core/tools/Get-IcingaCheckCommandConfig.psm1#L102-L115

Well, multiple things with this:
1. I don't understand why it's done this way in the first place instead of `"-NoProfile" = { set_if = true }`.
2. Also, it wouldn't be a problem with my suggestion. You could just do `{"-NoProfile" = true}` and would get the exact same result.
3. Yes, there are uses where this doesn't blow up. I mentioned it in my list because it can be used to build things that would cause trouble, think of something like where it's used to merge lists:

    ```
    "--exclude (global)" = {
      "key" = "--exclude"
      "value" = "$global_excludes$"
    }
    "--exclude (host)" = {
      "key" = "--exclude"
      "value" = "$host_excludes$"
    }
    ```

> Also there's no final decision yet whether we already need both value and set_if, see [Icinga/icinga-powershell-framework#633](https://github.com/Icinga/icinga-powershell-framework/pull/633) / "Things to consider while reviewing".

I think that might even be a reason for my `ifw_api_body` suggestion where there's just a DSL function that has full control over the JSON body. If IfW needed something slightly different, it could adjust it there. Also, we wouldn't hard-code any assumptions over the request JSON into the agent and IfW could even implement more fancy inputs without requiring an agent update.

Reply:
Regarding especially the latter: please not. It's not up to IfW, but up to the end users(' tools abilities) which of these to choose:

> 2\. Look at how the macro processor interprets set_if. You can set `true`, `"true"`, `"1"`, `1`, ...

Therefore **I insist**:

> Duplicating such magic logic is a bad idea in the long term. In contrast re-using both the code and the data structure LGTM.

And I don’t want to have to coordinate every single new thing in the future, nor to make yet another version no one upgrades (the whole cluster) to immediately. (Hello new capability!) Let's assume Christian approves my intended-inconsistencies strategy, but discovers an overseen 🐘 post factum. He won't have to supplicate anyone, but can just make a new IfW release by himself. (Yes, the lambda restriction can also turn out wrong, but lambdas can be wrapped in CVs, so it's OK.) Also Command#arguments is a well-known and speaking format, I see absolutely no reason for re-inventing any wheel here.

On the other hand we already allow only functions and scalars in Command#arguments, so I can restrict that check to scalars if you really want.